### PR TITLE
Runner - Support torch.distributed mode in runner

### PR DESCRIPTION
**Pull request by abuccts**
__Major Revisions__

* Support `torch.distributed` mode in runner.

### CODE REVIEW

File: superbench/runner/runner.py [hunk 1]
```diff
@@ -58,6 +58,27 @@ def __get_enabled_benchmarks(self):
...
+            exec_command (str): Executor command.
+
+        Return:
+            str: Runner command.
+        """
+        if mode == 'torch.distributed':
+            # TODO: replace with torch.distributed.run in v1.9
+            return (
+                'python3 -m torch.distributed.launch '
+                '--use_env --no_python --nproc_per_node=8 '
```
Comment (cp5555) [inline superbench/runner/runner.py:75]: Is it hard code for `--nproc_per_node`?

**Review by cp5555 — APPROVED** [review]

File: superbench/runner/runner.py [hunk 2]
```diff
@@ -58,6 +58,27 @@ def __get_enabled_benchmarks(self):
...
+        """Get runner command for given mode.
+
+        Args:
+            mode (str): Runner mode.
+            exec_command (str): Executor command.
+
+        Return:
+            str: Runner command.
+        """
+        if mode == 'torch.distributed':
```
Comment (guoshzhao) [inline superbench/runner/runner.py:71]: How to distinguish "distributed training" and "local parallel training per single node"?

Comment (abuccts) [reply superbench/runner/runner.py:71]: there's no need to distinguish? the only difference is `nnode=1` or `nnode>1`

File: superbench/runner/runner.py (hunk 1 above)
Comment (abuccts) [reply superbench/runner/runner.py:75]: currently yes, will change to `--nproc_per_node=gpu` in torch 1.9, here's [reference](https://github.com/pytorch/pytorch/blob/b4f3a989da375675785754e4b12d2ce757a1ffaa/torch/distributed/run.py#L269-L273)

if we want to support dynamic number, e.g. run distributed training on 4 gpus even if there're 8, we may need to add an option in config file.

Comment (cp5555) [reply superbench/runner/runner.py:75]: Yes, we definitely need to support it.

File: superbench/runner/runner.py [hunk 3]
```diff
@@ -58,6 +58,30 @@ def __get_enabled_benchmarks(self):
...
+            # TODO: replace with torch.distributed.run in v1.9
+            # TODO: only supports node_num=1 and node_num=all currently
+            return (
+                'python3 -m torch.distributed.launch '
+                '--use_env --no_python --nproc_per_node={proc_num} '
+                '--nnodes={node_num} --node_rank=$NODE_RANK '
+                '--master_addr=$MASTER_ADDR --master_port=$MASTER_PORT '
+                '{command}'
+            ).format(
+                proc_num=mode.proc_num or 8, node_num=1 if mode.node_num == 1 else '$NNODES', command=exec_command
```
Comment (guoshzhao) [inline superbench/runner/runner.py:81]: Who set $NODE_RANK, $NNODES, $MASTER_ADDR etc.? If node_num=1, master_addr should be `localhost`?

**Review by guoshzhao — APPROVED** [review]

File: superbench/runner/runner.py (hunk 3 above)
Comment (abuccts) [reply superbench/runner/runner.py:81]: it's set here:
https://github.com/microsoft/superbenchmark/blob/4063d8ebbd5760416f9e002514beccd070c5a483/superbench/runner/playbooks/check_env.yaml#L24-L28

File: superbench/runner/runner.py (hunk 1 above)
Comment (abuccts) [reply superbench/runner/runner.py]: added

File: superbench/runner/runner.py (hunk 2 above)
Comment (abuccts) [reply superbench/runner/runner.py]: added `proc_num` and `node_num` config

**Pull request merged**